reduce the amount of dependencies and other improvements

egonelbre opened the pull request:
This contains few commits to reduce the total number of dependencies significantly.

Most controversial change is probably removal of `github.com/npillmayer/schuko`. I tried to replace it with something reasonable and something that doesn't have any overhead -- usually. Currently this will require `uaxtrace` tag to be used to enable it.

I feel like pooling could be done in a better way, than in the current commit, but, if there are good benchmarks I can take a look and implement something better.

There's still github.com/cloudfoundry/jibber_jabber dependency, which is both archived and brings in ginkgo/gomega, without good reason.

egonelbre committed:
generators: remove arraylist

egonelbre committed:
remove go-commons-pool dependency

egonelbre committed:
clear UserData and preserve penalties cache

egonelbre committed:
remove github.com/npillmayer/schuko dependency

That dependency brings in a lot of other dependencies, which makes
integrating this into other projects more difficult.

Rewrote the tracing to use `log` package that's only included when
uaxtrace tag is used. For example:

  go test -tags uaxtrace -v ./...

whereswaldon commented:
See also #10 ; Thanks for taking a stab at this @egonelbre!

whereswaldon commented:
@egonelbre Your changes don't compile for me:

```
$ go build ./...
# github.com/npillmayer/uax/segment
segment/segment.go:403:24: cannot call pointer method Errorf on tracing.Trace
segment/segment.go:623:39: cannot call pointer method Debugf on tracing.Trace
# github.com/npillmayer/uax/uax29
uax29/uax29.go:166:25: cannot call pointer method Debugf on tracing.Trace
uax29/uax29.go:196:24: cannot call pointer method Debugf on tracing.Trace
uax29/uax29.go:198:24: cannot call pointer method Debugf on tracing.Trace
# github.com/npillmayer/uax/uax14
uax14/uax14.go:195:26: cannot call pointer method Debugf on tracing.Trace
uax14/uax14.go:202:26: cannot call pointer method Debugf on tracing.Trace
uax14/uax14.go:380:23: cannot call pointer method Debugf on tracing.Trace
```

The following patch fixes this, and I believe it's correct. Please let me know if you see issues with it.

```diff
diff --git a/internal/tracing/common.go b/internal/tracing/common.go
index 1b619c8..4985fff 100644
--- a/internal/tracing/common.go
+++ b/internal/tracing/common.go
@@ -2,7 +2,7 @@ package tracing

 var output Trace

-func P(key string, val interface{}) Trace       { return output.P(key, val) }
+func P(key string, val interface{}) *Trace      { return output.P(key, val) }
 func Debugf(format string, args ...interface{}) { output.Debugf(format, args) }
 func Infof(format string, args ...interface{})  { output.Infof(format, args) }
 func Errorf(format string, args ...interface{}) { output.Errorf(format, args) }
diff --git a/internal/tracing/notrace.go b/internal/tracing/notrace.go
index af1f3f7..ea856a8 100644
--- a/internal/tracing/notrace.go
+++ b/internal/tracing/notrace.go
@@ -7,7 +7,7 @@ type Trace struct{}

 func SetTestingLog(tb TB) {}

-func (trace *Trace) P(key string, val interface{}) Trace       { return Trace{} }
+func (trace *Trace) P(key string, val interface{}) *Trace      { return &Trace{} }
 func (trace *Trace) Debugf(format string, args ...interface{}) {}
 func (trace *Trace) Infof(format string, args ...interface{})  {}
 func (trace *Trace) Errorf(format string, args ...interface{}) {}
```

egonelbre committed:
fix few tracing calls

egonelbre commented:
@whereswaldon fixed. I'm avoiding using pointers and allocations, hence the slightly different fix.

egonelbre committed:
internal/testdata: unify ucd data

Currently UCD data was spread out across multiple packages.
Move the data into a single folder and add script for downloading
the data.

egonelbre committed:
internal/testdata: use in generators

This makes generators repeatable.

egonelbre committed:
uax14: rewrite the code generator

This way the range tables are much smaller.

egonelbre committed:
uax29: rewrite the code generator

egonelbre committed:
emoji: rewrite the code generator

egonelbre committed:
grapheme: rewrite code generator

It looks like the previous generator was not working properly.
Some runes cannot be printed with %+q, because they do not result in a valid
string. Hence the Control range table ended up with wrong set:

  fmt.Printf("%+q", 0xD800)
  // Outputs   \uFFFD
  // Expected  \uD800

The code should use %x or %U or %#U to avoid rune sanitization.

egonelbre committed:
uax14: move known failure handling

It's easier to update ucd data, if they aren't manually modified.

egonelbre committed:
shaping,internal/tablegen: output ranges in sorted order

egonelbre committed:
emoji: canonicalize generation

Changes to prepare for unifying table generation.

egonelbre committed:
grapheme: normalize class generator

egonelbre committed:
uax14: canonicalize class generator

egonelbre committed:
uax29: canonicalize class generation

egonelbre committed:
internal/classgen: unify generator

egonelbre committed:
internal/classgen: make output diffable

Output the range tables on separate lines, this makes diffs to updates
shorter and also the line lengths are smaller.

egonelbre committed:
shaping: use internal/classgen

Use classgen for generating the tables to reduce the amount
of code needed to maintain.

This also uses the v11 of Unicode data to ensure consistency
across packages.

egonelbre committed:
internal/tablegen: delete unused code

egonelbre commented:
Note the change has significantly grown to reduce the startup cost and the compiled binary size.

egonelbre committed:
segment: fix MaxInt64 usage

egonelbre committed:
grapheme,internal: fix formatting

egonelbre committed:
internal/testdata: avoid embedding

egonelbre committed:
bidi/internal/gen: simplify generator

egonelbre committed:
internal/testdata: update to v14

egonelbre committed:
uax11: remove ContextFromEnvironment

The jibber_jabber dependency causes a lot of additional dependencies,
it would be nicer to reimplemnt, however, we'll remove it for now.
